feat(workshop): add NoAgeRestrictions property and update age validation

synapsetron opened the pull request:
Introduce NoAgeRestrictions boolean to WorkshopMainRequiredPropertiesDto and WorkshopBaseDto. Make MinAge and MaxAge nullable and conditionally required. Update Validate method to handle NoAgeRestrictions logic. Ignore NoAgeRestrictions in WorkshopDraftMappingProfile and MappingProfile. Add unit tests for new validation logic.
Minor changes in ApplicationControllerTests and ChildControllerTests.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Workshops now offer flexible age restriction settings, including a new option to indicate no age restrictions, with automatic default boundaries applied when applicable.

- **Tests**
  - A new test class has been introduced to validate the behavior of the workshop model concerning age restrictions, ensuring robust validation of workshop configurations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

synapsetron committed:
feat(workshop): add NoAgeRestrictions property and update age validation

Introduce NoAgeRestrictions boolean to WorkshopMainRequiredPropertiesDto and WorkshopBaseDto.
Make MinAge and MaxAge nullable and conditionally required.
Update Validate method to handle NoAgeRestrictions logic.
Ignore NoAgeRestrictions in WorkshopDraftMappingProfile and MappingProfile.
Add unit tests for new validation logic.
Minor changes in ApplicationControllerTests and ChildControllerTests.

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Util/MappingProfile.cs:
```diff
@@ -187,7 +188,8 @@ public MappingProfile()
             .ForMember(dest => dest.Instagram,
-                opt => opt.MapFrom(src => src.Contacts.FirstOrDefault(c => c.IsDefault).SocialNetworks.FirstOrDefault(s => s.Type == SocialNetworkContactType.Instagram).Url));
+                opt => opt.MapFrom(src => src.Contacts.FirstOrDefault(c => c.IsDefault).SocialNetworks.FirstOrDefault(s => s.Type == SocialNetworkContactType.Instagram).Url))
+            .ForMember(dest => dest.NoAgeRestrictions, opt => opt.Ignore());
```
In hindsight, we probably should map it :) with logic like `if minage = 0 and maxage = 120 then true else false`

synapsetron replied:
Resolved

DmyMi commented on OutOfSchool/OutOfSchool.BusinessLogic/Models/Workshops/TempSave/WorkshopMainRequiredPropertiesDto.cs:
```diff
@@ -28,13 +29,16 @@ public class WorkshopMainRequiredPropertiesDto : IValidatableObject
-    [Required(ErrorMessage = "Children's min age is required")]
-    [Range(0, 120, ErrorMessage = "Min age should be a number from 0 to 120")]
-    public int MinAge { get; set; }
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingDefault)]
```
In hindsight, we probably should return it 😅

synapsetron replied:
done

synapsetron committed:
fix ( mapping) : added mapping to noAgeRestrictions to align with this logic:

NoAgeRestrictions = true — if there is no age restriction (0–120).

NoAgeRestrictions = false — if there are actual age restrictions.

fix ( WorkshopDTO) : deleted JSON ignore property on NoAgeRestirction field

synapsetron committed:
fix ( dto) : delete json ignore from dto
 fix ( mapping): add mapping to align logic with ageRestriction
